File: 1031-maximum-sum-of-two-non-overlapping-subarrays/1031-maximum-sum-of-two-non-overlapping-subarrays.js
/**
 * @param {number[]} nums
 * @param {number} firstLen
 * @param {number} secondLen
 * @return {number}
 */
var maxSumTwoNoOverlap = function(nums, firstLen, secondLen) {
  return Math.max(maxSum(nums, firstLen, secondLen), maxSum(nums, secondLen, firstLen));
};

const maxSum = (nums, lenL, lenM) => {
  let [sumL, sumM] = [0, 0];
  
  for (let i = 0; i < lenL + lenM; i++) 
    i < lenL ? sumL += nums[i] : sumM += nums[i];
  
  let [maxL, ans] = [sumL, sumL + sumM];
  
  for (let i = lenL + lenM; i < nums.length; i++) {
    sumL += nums[i - lenM] - nums[i - lenL - lenM];
    maxL = Math.max(maxL, sumL);
    sumM += nums[i] - nums[i - lenM];
    ans = Math.max(ans, maxL + sumM);
  }
  
  return ans;
}

File: 1031-maximum-sum-of-two-non-overlapping-subarrays/1031-maximum-sum-of-two-non-overlapping-subarrays.py
class Solution:
    def maxSumTwoNoOverlap(self, nums: List[int], firstLen: int, secondLen: int) -> int:
        def maxSum(nums, L, M):
            sum1, sum2 = 0, 0
            for i in range(L + M):
                if i < L:
                    sum1 += nums[i]
                else:
                    sum2 += nums[i]
            max_sum1, ans = sum1, sum1 + sum2
            for i in range(L+M, len(nums)):
                sum1 += nums[i - M] - nums[i - L - M]
                sum2 += nums[i] - nums[i - M]
                max_sum1 = max(max_sum1, sum1)
                ans = max(ans, max_sum1 + sum2)
            return ans
        
        
        return max(maxSum(nums, firstLen, secondLen), maxSum(nums, secondLen, firstLen))